keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1080 @@
+(for allocating claims) and the DRA drivers' kubelet plugin (for configuring
+the resource on a node). During scheduling of a pod, kube-scheduler and the DRA
+driver controller(s) that handle claims used by the pod communicate through the
+apiserver by updating the PodSchedulingContext object, ultimately leading to
+allocation of all pending claims and the pod being scheduled onto a node.
+
+This approach poses a problem for [Cluster
+Autoscaler](https://github.com/kubernetes/autoscaler) (CA) because it cannot
+simulate the effect of allocating or deallocating claims. "Numeric parameters"
+is an extension of DRA that addresses this by making claim parameters less
+opaque. "Numeric models" define how resource capacity per node can be published
+and the claim parameters that allocate some of that capacity for a claim.
+
+This KEP describes the core changes needed for this. Specific numeric models
+get defined in separate KEPs. Developers will be able to add their own numeric
+models by recompiling kube-scheduler and/or CA without having to modify any of
+the upstream code.
+
+When opting into using numeric parameters for their DRA driver, driver
+developers no longer need to provide their own controller. Kubernetes takes
+over allocating and deallocating claims.
+
+Scheduling performance is expected to become better compared to using the
+PodSchedulingContext. However, supporting CA is the main reason for this KEP.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+- Enable cluster autoscaling when pods use resource claims, with correct
+  decisions and changing the cluster size by more than one node at a time.
+
+- Support node-local resources. Adding or removing nodes has no effect
+  on network-attached resources and therefore CA does not need to (and cannot)
+  simulate them.
+
+- Allow DRA driver developers to provide a user experience that is similar to
+  the one possible without numeric parameters. Ideally, users should not notice
+  at all that a driver is using numeric parameters under the hood.
+
+- Support adding and versioning multiple different numeric models without
+  having to modify the implementation of DRA or the numeric parameters
+  extension.
+
+- Keep the core DRA functionality from [KEP
+  #3063](https://github.com/kubernetes/enhancements/issues/3063) working as
+  defined there for Kubernetes 1.29. Cluster autoscaling is not always needed
+  and in-tree numeric model(s) are not expected to cover all use case, so the
+  flexibility offered by opaque parameters remains relevant.
```

Comment:
Great question. You are right, I don't see this as a goal. From a "KEP goal" perspective, I think we should focus on ensuring the core use cases can be met while continuing to support cluster autoscaling. I think that an opaque "escape hatch" is a fallback / workaround and we should see how far we can go without it. In the proposal I have ("strongly typed" as Patrick describes it), we can go there if we need to - by adding a `custom` field that refers out to driver-specific data. But I would prefer to start by seeing how far we can get without that.

In other words, let's see if we can cover the 80% of use cases with a well-defined model with semantics that any client can rely on. For the other 20%, we can look at different workarounds. One is "add a custom model", but another, simpler one may be "just let them take the whole node". I am not convinced that there are use cases that need both the highly dynamic scheduling and the driver-specific logic.

My plan for the next day or two is to look through the use cases in detail and exercise the model to see if/how it may cover them.